Essentials/src/com/earth2me/essentials/signs/SignDisposal.java

```diff
@@ -12,8 +14,21 @@ public SignDisposal() {
     }
 
     @Override
-    protected boolean onSignInteract(final ISign sign, final User player, final String username, final IEssentials ess) {
-        player.getBase().openInventory(ess.getServer().createInventory(player.getBase(), 36, tl("disposal")));
+    protected boolean onSignInteract(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException, ChargeException {
+        if (!sign.getLine(3).isEmpty()) {
```

Comment:
It would probably actually be more user-friendly if this charge was only brought when the sign was actually used (an item placed inside).

This requires a special inventory handler specifically which will hook this sign's cost.

Reply:
There currently isn't any handling for disposal inventories whatsoever, and the only method I can think of for identifying them is to use the title, which isn't necessarily reliable (other plugins can use the same title and EssentialsX would interfere).

Reply:
It might be inefficient, but what about saving Player inventory state (like in a Map<IUser, Inventory>) and then operate depending on Player's presence/absence in the map?

Reply:
We could store a boolean like `hasDisposalOpen` on User and toggle this when they open/close the disposal inventory. I believe this approach is already used for `/invsee`?

Reply:
So the way I've always seen these types of things done (and the way I've always done them) is to create an 'essentials' InventoryHolder. You then extend this base EssInventoryHolder with specific disposal mechanics. I kind of lay it out here:

https://gist.github.com/Ichbinjoe/19b9fe88b91052ce4868fa3a33ac5fc1

Reply:
Alright, there's no trouble in creating InventoryHolder, but we will have no access to sign's `getTrade`

Reply:
So I didn't follow up on this, but the solution is to save the getTrade within the InventoryHolder implementation.

I think InventoryHolder is dead / being discouraged by Spigot now, so maybe a new solution has to be come up with.

Spigot!!!!